### DISCUSSION TITLE
How to have different allowed methods for different origins in CORSMiddleware?

### DISCUSSION THREAD

**Discussion by sohambhardwaj** [fastapi/fastapi#11794] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
origins = [
    "http://localhost.tiangolo.com",
    "https://localhost.tiangolo.com",
    "http://localhost",
    "http://localhost:8080",
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"], # Here I want different allowed methods for different origins
    allow_headers=["*"],
)
```

### Description

I want to have different allowed methods for different origins. For example for one url I want "POST" and "GET" methods to be allowed and for other I only want "GET" to be allowed. How can I achieve this? 

Will the below approach of writing a custom middleware work?
```
  class CustomCORSMiddleware(BaseHTTPMiddleware):
      async def dispatch(self, request: Request, call_next):
          response = await call_next(request)
          
          origin = request.headers.get('origin')
          
          if origin == "https://example1.com":
              response.headers["Access-Control-Allow-Origin"] = "https://example1.com"
              response.headers["Access-Control-Allow-Methods"] = "GET, POST"
          elif origin == "https://example2.com":
              response.headers["Access-Control-Allow-Origin"] = "https://example2.com"
              response.headers["Access-Control-Allow-Methods"] = "GET"          
          return response
  
  app.add_middleware(CustomCORSMiddleware)
```

### Operating System

Linux

### FastAPI Version

0.109.2

### Pydantic Version

2.8.0

### Python Version

Python 3.11.8

**Comment by YuriiMotov** [comment]
Looks like writing your middleware is the only way to achieve this.

Let's close this for now to keep the discussions clean.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
